File: simplified-books-core-ng/src/main/java/org/nypl/simplified/books/book_registry/BookStatusEvent.java
package org.nypl.simplified.books.book_registry;

import com.google.auto.value.AutoValue;

import org.nypl.simplified.books.book_database.BookEvent;
import org.nypl.simplified.books.book_database.BookID;

/**
 * The status of a book has changed.
 */

@AutoValue
public abstract class BookStatusEvent extends BookEvent {

  BookStatusEvent() {

  }

  /**
   * The type of status change.
   */

  public enum Type {

    /**
     * The book status changed.
     */

    BOOK_CHANGED,

    /**
     * The book status was removed.
     */

    BOOK_REMOVED
  }

  /**
   * @return The ID of the book in question
   */

  public abstract BookID book();

  /**
   * @return The type of status change
   */

  public abstract Type type();

  /**
   * Create a book status event.
   *
   * @param book The book
   * @param type The type
   * @return The event
   */

  public static BookStatusEvent create(
      final BookID book,
      final Type type) {
    return new AutoValue_BookStatusEvent(book, type);
  }
}

File: simplified-books-registry-api/src/main/java/org/nypl/simplified/books/book_registry/BookStatusEvent.kt
package org.nypl.simplified.books.book_registry

import org.nypl.simplified.books.api.BookEvent
import org.nypl.simplified.books.api.BookID

/**
 * The type of book status events.
 */

sealed class BookStatusEvent : BookEvent() {

  /**
   * @return The ID of the book in question
   */

  abstract val bookId: BookID

  /**
   * The previous book status.
   */

  abstract val statusPrevious: BookStatus?

  /**
   * The current book status.
   */

  abstract val statusNow: BookStatus?

  /**
   * A new book status was added to the registry.
   */

  data class BookStatusEventAdded(
    override val bookId: BookID,
    override val statusNow: BookStatus,
  ) : BookStatusEvent() {
    override val statusPrevious: BookStatus? =
      null
  }

  /**
   * The book status was removed from the registry.
   */

  data class BookStatusEventRemoved(
    override val bookId: BookID,
    override val statusPrevious: BookStatus,
  ) : BookStatusEvent() {
    override val statusNow: BookStatus? =
      null
  }

  /**
   * The book status was changed in the registry.
   */

  data class BookStatusEventChanged(
    override val bookId: BookID,
    override val statusPrevious: BookStatus?,
    override val statusNow: BookStatus
  ) : BookStatusEvent()
}
